net/veth.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	cleanup := func(format string, a ...interface{}) (*netlink.Veth, error) {
+		netlink.LinkDel(veth)
+		return nil, fmt.Errorf(format, a...)
+	}
+
+	switch maybeBridge.(type) {
+	case *netlink.Bridge:
+		if err := netlink.LinkSetMasterByIndex(veth, maybeBridge.Attrs().Index); err != nil {
+			return cleanup(`unable to set master of %s: %s`, name, err)
+		}
+	case *netlink.GenericLink:
+		if maybeBridge.Type() != "openvswitch" {
+			return cleanup(`device "%s" is of type "%s"`, bridgeName, maybeBridge.Type())
+		}
+		if err := odp.AddDatapathInterface(bridgeName, name); err != nil {
+			return cleanup(`failed to attach %s to device "%s": %s`, name, bridgeName, err)
+		}
+	case *netlink.Device:
+		// Assume it's our openvswitch device, and the kernel has not been updated to report the kind.
+		if err := odp.AddDatapathInterface(bridgeName, name); err != nil {
+			return cleanup(`failed to attach %s to device "%s": %s`, name, bridgeName, err)
+		}
+	default:
+		return cleanup(`device "%s" is not a bridge`, bridgeName)
+	}
+
+	if init != nil {
+		peer, err := netlink.LinkByName(peerName)
+		if err != nil {
+			return cleanup("unable to find peer veth %s: %s", peerName, err)
+		}
+		if err := init(peer); err != nil {
+			return cleanup("initializing veth: %s", err)
+		}
+	}
+
+	if err := netlink.LinkSetUp(veth); err != nil {
+		return cleanup("unable to bring veth up: %s", err)
+	}
+
+	return veth, nil
+}
+
+func AddAddresses(link netlink.Link, cidrs []*net.IPNet) (newAddrs []*net.IPNet, err error) {
+	existingAddrs, err := netlink.AddrList(link, netlink.FAMILY_V4)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get IP address for %q: %v", link.Attrs().Name, err)
+	}
+	for _, ipnet := range cidrs {
+		if contains(existingAddrs, ipnet) {
+			continue
+		}
+		if err := netlink.AddrAdd(link, &netlink.Addr{IPNet: ipnet}); err != nil {
+			return nil, fmt.Errorf("failed to add IP address to %q: %v", link.Attrs().Name, err)
+		}
+		newAddrs = append(newAddrs, ipnet)
+	}
+	return newAddrs, nil
+}
+
+func contains(addrs []netlink.Addr, addr *net.IPNet) bool {
+	for _, x := range addrs {
+		if addr.IP.Equal(x.IPNet.IP) {
+			return true
+		}
+	}
+	return false
+}
+
+const (
+	VethName         = "ethwe"        // name inside container namespace
+	vethPrefix       = "v" + VethName // starts with "veth" to suppress UI notifications
+	maxNetDevNameLen = 15
+)
+
+func AttachContainer(ns netns.NsHandle, id, ifName, bridgeName string, mtu int, withMulticastRoute bool, cidrs []*net.IPNet) error {
+	// Check if interface already exists
+	if err := common.WithNetNS(ns, func() error {
+		_, err := netlink.LinkByName(ifName)
+		return err
+	}); err != nil {
+		maxIDLen := maxNetDevNameLen - len(vethPrefix+"pl")
+		if len(id) > maxIDLen {
+			id = id[:maxIDLen] // trim passed ID if too long
+		}
+		name, peerName := vethPrefix+"pl"+id, vethPrefix+"pg"+id
+		_, err = CreateAndAttachVeth(name, peerName, bridgeName, mtu, func(veth netlink.Link) error {
```

Comment:
The old order in the `weave` script was:

create veth
ethtool tx off
move guest end inside netns
set name inside netns
set host end up
add host end to bridge
configure arp cache
add addresses
set guest end up
send gratuitous arp

The old order in the CNI plugin was:

create veth
add host end to bridge
move guest end inside netns
set host end up
set name inside netns
set guest end up
configure arp cache
add address

The new order is:

create veth
add host end to bridge
ethtool tx off
move guest end inside netns
set name inside netns
configure arp cache
set host end up
add addresses
set guest end up
send gratuitous arp

The previous differences between script and CNI indicate that the order doesn't matter too much, although CNI callers probably don't attempt to use the interface before the plugin returns after completing all tasks.

Most of the differences arise from sharing implementation with the Docker plugin, which cannot do the parts inside the container's netns.

Setting the host end up later seems like an improvement.